File: docs/moonwave/tokens.py
# SOURCE: https://github.com/evaera/moonwave/blob/master/docusaurus-plugin-moonwave/src/components/LuaType.js
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.

import re

def is_punc(char):
    return re.match(r'[\{\}<>\-\|&()\[\]]', char) is not None

def is_whitespace(char):
    return re.match(r'\s', char) is not None

def is_atom(char):
    return not is_whitespace(char) and not is_punc(char)

def tokenize(code, is_group=False):
    position = 0
    tokens = []
    
    def next_():
        nonlocal position
        if position < len(code):
            result = code[position]
        else:
            result = ""
        position += 1
        return result
    
    def peek():
        if position < len(code):
            return code[position]
        return ""
    
    def read(condition):
        buffer = ""
        
        while peek() and condition(peek()):
            buffer += next_()
        
        return buffer
    
    def read_balanced(left, right):
        buffer = ""
        depth = 0
        
        while peek():
            if peek() == left:
                depth += 1
            elif peek() == right:
                if depth == 0:
                    break
                else:
                    depth -= 1
            
            buffer += next_()
        
        return buffer
    
    
    while position < len(code):
        read(is_whitespace)
        
        if position >= len(code):
            break
        
        if peek() == "(":
            next_()
            tokens.append({
                "type": "tuple",
                "unseparated_tokens": tokenize(read_balanced("(", ")"), True),
            })
            next_()
            continue
        
        if peek() == "[":
            next_()
            tokens.append({
                "type": "indexer",
                "unseparated_tokens": tokenize(read_balanced("[", "]"), True),
            })
            next_()
            continue
        
        if peek() == "{":
            next_()
            tokens.append({
                "type": "table",
                "unseparated_tokens": tokenize(read_balanced("{", "}"), True),
            })
            next_()
            continue
        
        if (is_group and peek() == ","):
            next_()
            tokens.append({
                "type": "separator"
            })
            continue
        
        if is_punc(peek()):
            punc = next_()
            
            if punc == "-" and peek() == ">":
                tokens.append({
                    "type": "arrow"
                })
                next_()
                continue
            
            if punc == "|":
                tokens.append({"type": "union"})
                continue
            
            if punc == "&":
                tokens.append({"type": "intersection"})
                continue
            
            tokens.append({"type": "punc", "token": punc})
            continue
        
        atom = read(lambda char: char != "," and is_atom(char) if is_group else is_atom(char))
        
        if atom:
            if atom[-1] == ":":
                tokens.append({
                    "type": "identifier", 
                    "identifier": atom[:-1],
                })
            elif ":" in atom:
                identifier, type_ = atom.split(":", 1)
                tokens.append({
                    "type": "identifier",
                    "identifier": identifier
                })
                tokens.append({
                    "type": "lua_type",
                    "lua_type": type_
                })
            else:
                tokens.append({
                    "type": "lua_type", 
                    "lua_type": atom
                })
            continue
        
        raise ValueError(f"Reached bottom of tokenizer with no match: {peek()}")
        
        
    return tokens
    

File: docs/tokens.ts
const is_punc = (char: string): boolean => /[\{\}<>\-\|&()\[\]]/.test(char)
const is_whitespace = (char: string): boolean => /\s/.test(char)
const is_atom = (char: string): boolean => !(is_whitespace(char) || is_punc(char))

interface Token {
	type: string;
	identifier?: string;
	lua_type?: string;
	unseparated_tokens?: Token[];
	token?: string; // for punctuation
	separator?: true; // for comma separator
	arrow?: true; // for -> arrow
	union?: true; // for union operator
	intersection?: true; // for intersection operator
}

export function tokenize(code: string, is_group: boolean = false): Token[] {
	let position = 0
	let tokens: Token[] = []
	
	const peek = () => position < code.length ? code[position] : null
	const next = () => {
		let result = peek();
		position += 1
		return result
	}
	const read = (condition: (char: string) => boolean): string => {
		let buffer = ""
		while (peek() && condition(peek())) {buffer += next()}
		return buffer
	}
	const read_balanced = (left: string, right: string): string => {
		let buffer = ""
		let depth = 0
		while (peek()) {
			if (peek() == left) {depth += 1} else 
			if (peek() == right) {if (depth == 0) {break} else {depth -= 1}}
			buffer += next()
		}
		return buffer
	}
	
	const read_brackets = (v_type: string, left: string, right: string) => {
		next()
		tokens.push({
			type: v_type,
			unseparated_tokens: tokenize(read_balanced(left, right), true),
		})
		next()
	}
	
	while (position < code.length) {
		read(is_whitespace)
		
		if (position >= code.length) {break;}
		
		if (peek() == "(") {read_brackets("tuple", "(", ")"); continue;}
		if (peek() == "[") {read_brackets("indexer", "[", "]"); continue;}
		if (peek() == "{") {read_brackets("table", "{", "}"); continue;}
		if (is_group && peek() == ",") {
			next();
			tokens.push({type: "separator"});
			continue;
		}
		if (is_punc(peek())) {
			const punc = next();
			if (punc == "-" && peek() == ">") {
				tokens.push({type: "arrow"});
				next();
				continue
			}
			if (punc == "|") {tokens.push({type: "union"}); continue;}
			if (punc == "&") {tokens.push({type: "intersection"}); continue;}
			
			tokens.push({type: "punc", token: punc});
			continue;
		}
		
		
		const atom = read((char) => is_atom(char) && (char !== "," || !is_group))
		
		if (atom != "") {
			if (atom.endsWith(":")) {
				tokens.push({
					type: "identifier", 
					identifier: atom.slice(0, -1),
				})
			} else if (atom.includes(":")) {
				const [identifier, type_] = atom.split(":", 1)
				tokens.push({
					type: "identifier",
					identifier: identifier
				})
				tokens.push({
					type: "lua_type",
					lua_type: type_
				})
			} else {
				tokens.push({
					type: "lua_type", 
					lua_type: atom
				})
			}
			continue
		}
			
		
		throw new Error(`Reached bottom of tokenizer with no match: ${peek()} \"${atom}\" ${is_group} `)
	}
		
	return tokens
}